File: CD-Trabalho-6 javacript/shared/repository.py
"""
Repositório em memória - versão simplificada
"""
from typing import Optional, List, Dict
from shared.models import Usuario, Musica, Playlist, gerar_id


class Repositorio:
    """Repositório em memória para gerenciar operações CRUD"""
    
    def __init__(self):
        # Armazenamento em memória usando dicionários
        self._usuarios: Dict[str, Usuario] = {}
        self._musicas: Dict[str, Musica] = {}
        self._playlists: Dict[str, Playlist] = {}
    
    # ========== USUÁRIOS ==========
    
    def criar_usuario(self, nome: str, idade: int, id: Optional[str] = None) -> Usuario:
        """Cria um novo usuário"""
        usuario = Usuario(
            id=id or gerar_id(),
            nome=nome,
            idade=idade
        )
        self._usuarios[usuario.id] = usuario
        return usuario
    
    def obter_usuario(self, id: str) -> Optional[Usuario]:
        """Obtém um usuário por ID"""
        return self._usuarios.get(id)
    
    def listar_usuarios(self) -> List[Usuario]:
        """Lista todos os usuários"""
        return list(self._usuarios.values())
    
    def atualizar_usuario(self, id: str, nome: Optional[str] = None, idade: Optional[int] = None) -> Optional[Usuario]:
        """Atualiza um usuário"""
        usuario = self._usuarios.get(id)
        if not usuario:
            return None
        
        if nome is not None:
            usuario.nome = nome
        if idade is not None:
            usuario.idade = idade
        
        return usuario
    
    def remover_usuario(self, id: str) -> bool:
        """Remove um usuário e todas suas playlists"""
        if id not in self._usuarios:
            return False
        
        # Remove todas as playlists do usuário
        playlists_para_remover = [p.id for p in self._playlists.values() if p.usuario_id == id]
        for playlist_id in playlists_para_remover:
            del self._playlists[playlist_id]
        
        del self._usuarios[id]
        return True
    
    # ========== MÚSICAS ==========
    
    def criar_musica(self, nome: str, artista: str, id: Optional[str] = None) -> Musica:
        """Cria uma nova música"""
        musica = Musica(
            id=id or gerar_id(),
            nome=nome,
            artista=artista
        )
        self._musicas[musica.id] = musica
        return musica
    
    def obter_musica(self, id: str) -> Optional[Musica]:
        """Obtém uma música por ID"""
        return self._musicas.get(id)
    
    def listar_musicas(self) -> List[Musica]:
        """Lista todas as músicas"""
        return list(self._musicas.values())
    
    def atualizar_musica(self, id: str, nome: Optional[str] = None, artista: Optional[str] = None) -> Optional[Musica]:
        """Atualiza uma música"""
        musica = self._musicas.get(id)
        if not musica:
            return None
        
        if nome is not None:
            musica.nome = nome
        if artista is not None:
            musica.artista = artista
        
        return musica
    
    def remover_musica(self, id: str) -> bool:
        """Remove uma música de todas as playlists"""
        if id not in self._musicas:
            return False
        
        # Remove a música de todas as playlists
        for playlist in self._playlists.values():
            if id in playlist.musicas_ids:
                playlist.musicas_ids.remove(id)
        
        del self._musicas[id]
        return True
    
    # ========== PLAYLISTS ==========
    
    def criar_playlist(self, nome: str, usuario_id: str, musicas_ids: Optional[List[str]] = None, id: Optional[str] = None) -> Playlist:
        """Cria uma nova playlist"""
        # Valida se o usuário existe
        if usuario_id not in self._usuarios:
            raise ValueError('Usuário não encontrado')
        
        # Valida se todas as músicas existem
        if musicas_ids:
            for musica_id in musicas_ids:
                if musica_id not in self._musicas:
                    raise ValueError(f'Música {musica_id} não encontrada')
        
        playlist = Playlist(
            id=id or gerar_id(),
            nome=nome,
            usuario_id=usuario_id,
            musicas_ids=musicas_ids or []
        )
        self._playlists[playlist.id] = playlist
        return playlist
    
    def obter_playlist(self, id: str) -> Optional[Playlist]:
        """Obtém uma playlist por ID"""
        return self._playlists.get(id)
    
    def listar_playlists(self) -> List[Playlist]:
        """Lista todas as playlists"""
        return list(self._playlists.values())
    
    def listar_playlists_por_usuario(self, usuario_id: str) -> List[Playlist]:
        """Lista playlists de um usuário"""
        return [p for p in self._playlists.values() if p.usuario_id == usuario_id]
    
    def listar_musicas_por_playlist(self, playlist_id: str) -> List[Musica]:
        """Lista músicas de uma playlist"""
        playlist = self._playlists.get(playlist_id)
        if not playlist:
            return []
        
        musicas = []
        for musica_id in playlist.musicas_ids:
            musica = self._musicas.get(musica_id)
            if musica:
                musicas.append(musica)
        return musicas
    
    def listar_playlists_por_musica(self, musica_id: str) -> List[Playlist]:
        """Lista playlists que contêm uma música"""
        return [p for p in self._playlists.values() if musica_id in p.musicas_ids]
    
    def atualizar_playlist(self, id: str, nome: Optional[str] = None, usuario_id: Optional[str] = None) -> Optional[Playlist]:
        """Atualiza uma playlist"""
        playlist = self._playlists.get(id)
        if not playlist:
            return None
        
        # Valida usuário se estiver sendo alterado
        if usuario_id is not None:
            if usuario_id not in self._usuarios:
                raise ValueError('Usuário não encontrado')
            playlist.usuario_id = usuario_id
        
        if nome is not None:
            playlist.nome = nome
        
        return playlist
    
    def adicionar_musica_a_playlist(self, playlist_id: str, musica_id: str) -> Optional[Playlist]:
        """Adiciona uma música a uma playlist"""
        playlist = self._playlists.get(playlist_id)
        if not playlist:
            return None
        
        if musica_id not in self._musicas:
            raise ValueError('Música não encontrada')
        
        # Evita duplicatas
        if musica_id not in playlist.musicas_ids:
            playlist.musicas_ids.append(musica_id)
        
        return playlist
    
    def remover_musica_de_playlist(self, playlist_id: str, musica_id: str) -> Optional[Playlist]:
        """Remove uma música de uma playlist"""
        playlist = self._playlists.get(playlist_id)
        if not playlist:
            return None
        
        if musica_id in playlist.musicas_ids:
            playlist.musicas_ids.remove(musica_id)
        
        return playlist
    
    def remover_playlist(self, id: str) -> bool:
        """Remove uma playlist"""
        if id not in self._playlists:
            return False
        
        del self._playlists[id]
        return True


# Instância global compartilhada entre todos os serviços
_repositorio_global = Repositorio()


def get_repositorio() -> Repositorio:
    """Retorna a instância global do repositório"""
    return _repositorio_global

File: CD-Trabalho-6 javacript/shared/repository.js
/**
 * Repositório usando PostgreSQL - versão Node.js
 */
const { query } = require('./database');
const { Usuario, Musica, Playlist } = require('./models');
const { v4: uuidv4 } = require('uuid');

class Repositorio {
    // ========== USUÁRIOS ==========
    
    async criarUsuario(nome, idade, id = null) {
        const usuarioId = id || uuidv4();
        await query(
            'INSERT INTO usuarios (id, nome, idade) VALUES ($1, $2, $3)',
            [usuarioId, nome, idade]
        );
        return new Usuario(usuarioId, nome, idade);
    }
    
    async obterUsuario(id) {
        const result = await query('SELECT * FROM usuarios WHERE id = $1', [id]);
        if (result.rows.length === 0) {
            return null;
        }
        const row = result.rows[0];
        return new Usuario(row.id, row.nome, row.idade);
    }
    
    async listarUsuarios() {
        const result = await query('SELECT * FROM usuarios ORDER BY nome');
        return result.rows.map(row => new Usuario(row.id, row.nome, row.idade));
    }
    
    async atualizarUsuario(id, nome = null, idade = null) {
        const usuario = await this.obterUsuario(id);
        if (!usuario) {
            return null;
        }
        
        const updates = [];
        const values = [];
        let paramIndex = 1;
        
        if (nome !== null) {
            updates.push(`nome = $${paramIndex++}`);
            values.push(nome);
        }
        if (idade !== null) {
            updates.push(`idade = $${paramIndex++}`);
            values.push(idade);
        }
        
        if (updates.length === 0) {
            return usuario;
        }
        
        values.push(id);
        await query(
            `UPDATE usuarios SET ${updates.join(', ')} WHERE id = $${paramIndex}`,
            values
        );
        
        return await this.obterUsuario(id);
    }
    
    async removerUsuario(id) {
        const result = await query('DELETE FROM usuarios WHERE id = $1', [id]);
        return result.rowCount > 0;
    }
    
    // ========== MÚSICAS ==========
    
    async criarMusica(nome, artista, id = null) {
        const musicaId = id || uuidv4();
        await query(
            'INSERT INTO musicas (id, nome, artista) VALUES ($1, $2, $3)',
            [musicaId, nome, artista]
        );
        return new Musica(musicaId, nome, artista);
    }
    
    async obterMusica(id) {
        const result = await query('SELECT * FROM musicas WHERE id = $1', [id]);
        if (result.rows.length === 0) {
            return null;
        }
        const row = result.rows[0];
        return new Musica(row.id, row.nome, row.artista);
    }
    
    async listarMusicas() {
        const result = await query('SELECT * FROM musicas ORDER BY nome');
        return result.rows.map(row => new Musica(row.id, row.nome, row.artista));
    }
    
    async atualizarMusica(id, nome = null, artista = null) {
        const musica = await this.obterMusica(id);
        if (!musica) {
            return null;
        }
        
        const updates = [];
        const values = [];
        let paramIndex = 1;
        
        if (nome !== null) {
            updates.push(`nome = $${paramIndex++}`);
            values.push(nome);
        }
        if (artista !== null) {
            updates.push(`artista = $${paramIndex++}`);
            values.push(artista);
        }
        
        if (updates.length === 0) {
            return musica;
        }
        
        values.push(id);
        await query(
            `UPDATE musicas SET ${updates.join(', ')} WHERE id = $${paramIndex}`,
            values
        );
        
        return await this.obterMusica(id);
    }
    
    async removerMusica(id) {
        const result = await query('DELETE FROM musicas WHERE id = $1', [id]);
        return result.rowCount > 0;
    }
    
    // ========== PLAYLISTS ==========
    
    async criarPlaylist(nome, usuarioId, musicasIds = [], id = null) {
        // Valida se o usuário existe
        const usuario = await this.obterUsuario(usuarioId);
        if (!usuario) {
            throw new Error('Usuário não encontrado');
        }
        
        const playlistId = id || uuidv4();
        
        // Cria a playlist
        await query(
            'INSERT INTO playlists (id, nome, usuario_id) VALUES ($1, $2, $3)',
            [playlistId, nome, usuarioId]
        );
        
        // Adiciona músicas se fornecidas
        if (musicasIds && musicasIds.length > 0) {
            // Valida se todas as músicas existem
            const placeholders = musicasIds.map((_, i) => `$${i + 1}`).join(',');
            const musicasResult = await query(
                `SELECT id FROM musicas WHERE id IN (${placeholders})`,
                musicasIds
            );
            
            if (musicasResult.rows.length !== musicasIds.length) {
                // Remove a playlist criada
                await query('DELETE FROM playlists WHERE id = $1', [playlistId]);
                throw new Error('Uma ou mais músicas não foram encontradas');
            }
            
            // Insere as associações
            for (const musicaId of musicasIds) {
                await query(
                    'INSERT INTO playlist_musica (playlist_id, musica_id) VALUES ($1, $2) ON CONFLICT DO NOTHING',
                    [playlistId, musicaId]
                );
            }
        }
        
        return new Playlist(playlistId, nome, usuarioId, musicasIds);
    }
    
    async obterPlaylist(id) {
        const result = await query('SELECT * FROM playlists WHERE id = $1', [id]);
        if (result.rows.length === 0) {
            return null;
        }
        const row = result.rows[0];
        
        // Busca as músicas da playlist
        const musicasResult = await query(
            'SELECT musica_id FROM playlist_musica WHERE playlist_id = $1',
            [id]
        );
        const musicasIds = musicasResult.rows.map(r => r.musica_id);
        
        return new Playlist(row.id, row.nome, row.usuario_id, musicasIds);
    }
    
    async listarPlaylists() {
        try {
            // Otimização: busca todas as playlists e suas músicas em 2 queries ao invés de N+1
            const result = await query('SELECT * FROM playlists ORDER BY nome');
            
            if (result.rows.length === 0) {
                return [];
            }
            
            // Busca todas as associações de uma vez usando IN
            // PostgreSQL tem limite de parâmetros, então processa em lotes se necessário
            const playlistIds = result.rows.map(r => r.id);
            const musicasPorPlaylist = {};
            
            // Processa em lotes de 1000 para evitar limite de parâmetros
            const BATCH_SIZE = 1000;
            for (let i = 0; i < playlistIds.length; i += BATCH_SIZE) {
                const batch = playlistIds.slice(i, i + BATCH_SIZE);
                const placeholders = batch.map((_, idx) => `$${idx + 1}`).join(',');
                const musicasResult = await query(
                    `SELECT playlist_id, musica_id FROM playlist_musica WHERE playlist_id IN (${placeholders})`,
                    batch
                );
                
                // Agrupa músicas por playlist
                for (const row of musicasResult.rows) {
                    if (!musicasPorPlaylist[row.playlist_id]) {
                        musicasPorPlaylist[row.playlist_id] = [];
                    }
                    musicasPorPlaylist[row.playlist_id].push(row.musica_id);
                }
            }
            
            // Cria as playlists com suas músicas
            const playlists = [];
            for (const row of result.rows) {
                const musicasIds = musicasPorPlaylist[row.id] || [];
                playlists.push(new Playlist(row.id, row.nome, row.usuario_id, musicasIds));
            }
            
            return playlists;
        } catch (error) {
            console.error('Erro em listarPlaylists:', error);
            throw error;
        }
    }
    
    async listarPlaylistsPorUsuario(usuarioId) {
        try {
            // Otimização: busca todas as playlists e suas músicas em 2 queries ao invés de N+1
            const result = await query(
                'SELECT * FROM playlists WHERE usuario_id = $1 ORDER BY nome',
                [usuarioId]
            );
            
            if (result.rows.length === 0) {
                return [];
            }
            
            // Busca todas as associações de uma vez usando IN
            const playlistIds = result.rows.map(r => r.id);
            if (playlistIds.length === 0) {
                return result.rows.map(row => new Playlist(row.id, row.nome, row.usuario_id, []));
            }
            
            const placeholders = playlistIds.map((_, i) => `$${i + 1}`).join(',');
            const musicasResult = await query(
                `SELECT playlist_id, musica_id FROM playlist_musica WHERE playlist_id IN (${placeholders})`,
                playlistIds
            );
            
            // Agrupa músicas por playlist
            const musicasPorPlaylist = {};
            for (const row of musicasResult.rows) {
                if (!musicasPorPlaylist[row.playlist_id]) {
                    musicasPorPlaylist[row.playlist_id] = [];
                }
                musicasPorPlaylist[row.playlist_id].push(row.musica_id);
            }
            
            // Cria as playlists com suas músicas
            const playlists = [];
            for (const row of result.rows) {
                const musicasIds = musicasPorPlaylist[row.id] || [];
                playlists.push(new Playlist(row.id, row.nome, row.usuario_id, musicasIds));
            }
            
            return playlists;
        } catch (error) {
            console.error('Erro em listarPlaylistsPorUsuario:', error);
            throw error;
        }
    }
    
    async listarMusicasPorPlaylist(playlistId) {
        const result = await query(
            `SELECT m.* FROM musicas m
             INNER JOIN playlist_musica pm ON m.id = pm.musica_id
             WHERE pm.playlist_id = $1
             ORDER BY m.nome`,
            [playlistId]
        );
        return result.rows.map(row => new Musica(row.id, row.nome, row.artista));
    }
    
    async listarPlaylistsPorMusica(musicaId) {
        try {
            // Otimização: busca todas as playlists e suas músicas em 2 queries ao invés de N+1
            const result = await query(
                `SELECT p.* FROM playlists p
                 INNER JOIN playlist_musica pm ON p.id = pm.playlist_id
                 WHERE pm.musica_id = $1
                 ORDER BY p.nome`,
                [musicaId]
            );
            
            if (result.rows.length === 0) {
                return [];
            }
            
            // Busca todas as associações de uma vez usando IN
            const playlistIds = result.rows.map(r => r.id);
            if (playlistIds.length === 0) {
                return result.rows.map(row => new Playlist(row.id, row.nome, row.usuario_id, []));
            }
            
            const placeholders = playlistIds.map((_, i) => `$${i + 1}`).join(',');
            const musicasResult = await query(
                `SELECT playlist_id, musica_id FROM playlist_musica WHERE playlist_id IN (${placeholders})`,
                playlistIds
            );
            
            // Agrupa músicas por playlist
            const musicasPorPlaylist = {};
            for (const row of musicasResult.rows) {
                if (!musicasPorPlaylist[row.playlist_id]) {
                    musicasPorPlaylist[row.playlist_id] = [];
                }
                musicasPorPlaylist[row.playlist_id].push(row.musica_id);
            }
            
            // Cria as playlists com suas músicas
            const playlists = [];
            for (const row of result.rows) {
                const musicasIds = musicasPorPlaylist[row.id] || [];
                playlists.push(new Playlist(row.id, row.nome, row.usuario_id, musicasIds));
            }
            
            return playlists;
        } catch (error) {
            console.error('Erro em listarPlaylistsPorMusica:', error);
            throw error;
        }
    }
    
    async atualizarPlaylist(id, nome = null, usuarioId = null) {
        const playlist = await this.obterPlaylist(id);
        if (!playlist) {
            return null;
        }
        
        const updates = [];
        const values = [];
        let paramIndex = 1;
        
        if (nome !== null) {
            updates.push(`nome = $${paramIndex++}`);
            values.push(nome);
        }
        if (usuarioId !== null) {
            // Valida se o usuário existe
            const usuario = await this.obterUsuario(usuarioId);
            if (!usuario) {
                throw new Error('Usuário não encontrado');
            }
            updates.push(`usuario_id = $${paramIndex++}`);
            values.push(usuarioId);
        }
        
        if (updates.length === 0) {
            return playlist;
        }
        
        values.push(id);
        await query(
            `UPDATE playlists SET ${updates.join(', ')} WHERE id = $${paramIndex}`,
            values
        );
        
        return await this.obterPlaylist(id);
    }
    
    async adicionarMusicaAPlaylist(playlistId, musicaId) {
        const playlist = await this.obterPlaylist(playlistId);
        if (!playlist) {
            return null;
        }
        
        const musica = await this.obterMusica(musicaId);
        if (!musica) {
            throw new Error('Música não encontrada');
        }
        
        // Evita duplicatas
        if (!playlist.musicasIds.includes(musicaId)) {
            await query(
                'INSERT INTO playlist_musica (playlist_id, musica_id) VALUES ($1, $2) ON CONFLICT DO NOTHING',
                [playlistId, musicaId]
            );
        }
        
        return await this.obterPlaylist(playlistId);
    }
    
    async removerMusicaDePlaylist(playlistId, musicaId) {
        const playlist = await this.obterPlaylist(playlistId);
        if (!playlist) {
            return null;
        }
        
        await query(
            'DELETE FROM playlist_musica WHERE playlist_id = $1 AND musica_id = $2',
            [playlistId, musicaId]
        );
        
        return await this.obterPlaylist(playlistId);
    }
    
    async removerPlaylist(id) {
        const result = await query('DELETE FROM playlists WHERE id = $1', [id]);
        return result.rowCount > 0;
    }
}

// Instância global compartilhada entre todos os serviços
const _repositorioGlobal = new Repositorio();

function getRepositorio() {
    return _repositorioGlobal;
}

module.exports = {
    Repositorio,
    getRepositorio
};
